File: bot.js
/**************************************************************************
                      ALANIS MORISSETTE DISCORD BOT
                      - Posts Alanis Morissette lyrics
                      - Calls Al gay
                      - Pings the hell out of Zerahl
                      - Sends thots and prayers
                      Written By: @Indoctrine#1337
                      Contributors: @Zerahl#6487
**************************************************************************/
const Discord = require('discord.js');
const auth = require('./auth.json');
const fs = require('fs');
const ironic = require('./ironic.json');
const bot = new Discord.Client();
const path = "/var/www/html/alanisbot/";
var hacktriggers;
var hackattempts = 0;
var hackdifficulty = 10;    // 0=guaranteed(1attempt) 1=testing(<10attempts) 5=easy(~25attempts) 10=normal(~35attempts) 20=hard(~50attempts) 40=TheGibson(~70attempts) (d100 is 50% chance with ~70 attempts)

function readFileSync(filePath) {
    var options = {encoding:'utf-8', flag:'r'};
    var buffer = fs.readFileSync(path + filePath, options);
    return buffer;
}

function writeFileSync(filePath, content, flag = 'w') {
    var options = {encoding:'utf-8', flag: flag};
    fs.writeFileSync(path + filePath, content, options);
}

bot.on('ready', () => {
  bot.user.setActivity('Ironic - Alanis Morissette', {type: 'LISTENING'});
  console.log('Connected as ' + bot.user.tag);
});

bot.on('message', (message) => {
  if(message.author != bot.user){
    if(message.content.match(/[i|l1]\s*r\s*[o0]\s*n\s*[yi|l1]/i)){
      var randomphrase = Math.floor(Math.random() * Math.floor(ironic.phrases.length));
      message.channel.send(ironic.phrases[randomphrase]);
    }

    else if(message.content.match(/(big gay)/i)){
      message.channel.send('<@268862954919821323> is the biggest of gays.');
    }

    else if(message.content.match(/(:<)/i)){
      hackattempts = readFileSync('db/hackattempts.txt');
      hackattempts++;
      writeFileSync('db/hackattempts.txt', hackattempts);

      var randompercentage = Math.floor(100 * Math.pow((Math.random()), (hackdifficulty/hackattempts)) + 1);
      if(randompercentage == 100){
        hacktriggers = readFileSync('db/hacktriggers.txt');
        hacktriggers++;
        writeFileSync('db/hacktriggers.txt', hacktriggers);
        console.log('Hack trigger number ' + hacktriggers + ' triggered in ' + hackattempts + ' tries.');
        message.channel.send('Attempt 0x' + hackattempts + ' successful. Hacking ' + randompercentage + '% complete. Consequences will never be the same, <@64909909052887040> has backtraced your IP and the cyberpolice are on their way.');
        hackattempts = 0;
        writeFileSync('db/hackattempts.txt', hackattempts);
      }

      else{
        message.channel.send('<@64909909052887040> is a big nerd. Hacking in progress, ' + randompercentage + '% complete.');
      }
    }

    else if(message.content.match(/(!hax)/i)){
      hacktriggers = readFileSync('db/hacktriggers.txt');
      message.channel.send('<@64909909052887040> has hacked the Gibson ' + hacktriggers + ' times. <:gotem:548663899466235915>');
    }

    else if(message.content.match(/(thot)/i)){
      message.react('🙏');
      message.channel.send('Sending thots and prayers to <@' + message.author.id + '>');
    }
  }
});

bot.login(auth.token);
bot.on('disconnect', function(errMsg, code) {
  bot.login(auth.token);
});
bot.on('error', function(console_error){
  writeFileSync('log/errors.log', '\n' + console_error, 'a');
  console.log(console_error);
  bot.login(auth.token);
});

File: bot.py
import discord
from discord.ext import commands
import logging
import json
import sys
import random
import aiofiles
import asyncio
import math
import re

logging.basicConfig(format='%(asctime)s [%(levelname)s] %(message)s', stream=sys.stderr,level=logging.INFO)

async def load_dependencies():
    async with aiofiles.open('ironic.json', 'r') as f:
        ironic_contents = await f.read()
    async with aiofiles.open('hack_command.json', 'r') as g:
        hack_contents = await g.read()
    async with aiofiles.open('auth.json', 'r') as h:
        creds = await h.read()
    return json.loads(ironic_contents), json.loads(hack_contents), json.loads(creds)

intents = discord.Intents.default()
intents.members = True # Intent allows us to get users that haven't been seen yet
bot = commands.Bot(command_prefix='!', case_insensitive=True, intents=intents)
bot.ironic_statements = []
bot.ironic_statements, bot.hack_data, bot.creds = asyncio.get_event_loop().run_until_complete(load_dependencies())
extensions = ['cogs.hack']

if __name__ == '__main__':
    for extension in extensions:
        bot.load_extension(extension)

@bot.event
async def on_ready():
    await bot.change_presence(activity=discord.Activity(type=discord.ActivityType.listening, name='Ironic - Alanis Morissette', url='https://open.spotify.com/track/29YBihzQOmat0U74k4ukdx?si=RNDU7bMzSMKRzGYwnyqA-g'))
    logging.info(f'Logged on as {bot.user}!')

@bot.event
async def on_message(message):
    ironic_regex = re.compile(r'[i|l1]\s*r\s*[o0]\s*n\s*[yi|l1]', re.IGNORECASE)
    ctx = message.channel
    if message.author.id is bot.user.id:
        return
    if ':<' in message.content:
        bot.hack_data['attempts'] += 1
        randompercentage = math.floor(100 * math.pow((random.random()), (bot.hack_data['difficulty']/bot.hack_data['attempts'])))
        if(randompercentage == 100):
            bot.hack_data['triggers'] += 1
            await ctx.send(f"Attempt 0x{bot.hack_data['attempts']} successful. Hacking {randompercentage}% complete. Consequences will never be the same, <@!64909909052887040> has backtraced your IP and the cyberpolice are on their way.");
            bot.hack_data['attempts'] = 0
        else:
            if randompercentage > 60:
                await ctx.send(f'<@!64909909052887040> is getting closer... Hacking {randompercentage}% complete.');
            else:
                await ctx.send(f'<@!64909909052887040> is a big nerd. Hacking in progress, {randompercentage}% complete.')
        async with aiofiles.open('hack_command.json', 'w') as f:
            string_json = json.dumps(bot.hack_data)
            await f.write(string_json)
    elif ironic_regex.search(message.content):
        await ctx.send(random.choice(bot.ironic_statements))
    elif re.search('thot', message.content, flags=re.IGNORECASE):
        await message.add_reaction('🙏')
        await ctx.send(f'Sending thots and prayers to <@!{message.author.id}>')
    await bot.process_commands(message)

try:
    bot.run(bot.creds['token'])
except Exception as e:
    logging.exception(f'Could not connect to Discord {e}')
    sys.exit(1)
